### PR TITLE
[Bugfix] Respect min_tokens in scheduler stop check

### PR DESCRIPTION
**Pull request by elaineyz**
## Purpose
Fixes a bug where in the scheduler's stop_check logic, generation can stop on EOS token before min_tokens requirements have been met.

## Test Plan
Added corresponding unit tests. Also tested manually E2E on Neuron and verified the expected behavior. 

## Test Result
Ran the same prompt + sampling_params on a TinyLlama model (EOS=2) on Neuron:

sampling_params = SamplingParams(top_k=10, temperature=0.8, top_p=0.95, min_tokens=32, max_tokens=64)

`without fix` -> stopped on first output token (EOS token)

Prompt: 'explain the fractional banking thoroughly.'
Generated text: ''
output: RequestOutput(request_id=0, prompt='explain the fractional banking thoroughly.', prompt_token_ids=[1, 5649, 278, 15958, 284, 9124, 292, 26606, 29889], encoder_prompt=None, encoder_prompt_token_ids=None, prompt_logprobs=None, outputs=[CompletionOutput(index=0, text='', token_ids=[2], cumulative_logprob=None, logprobs=None, finish_reason=stop, stop_reason=None)], finished=True, metrics=None, lora_request=None, num_cached_tokens=0, multi_modal_placeholders={})

`with fix` -> continues generation after 1st EOS token since min_tokens have not been met 

Prompt: 'explain the fractional banking thoroughly.'
Generated text: ' \n<|assistant|>\nFractional banking is a system of banking that allows for the creation of fractional reserve banking, where the bank creates more money than is needed to meet the demands of the economy. In this system, banks create new money out of thin air by'
output: RequestOutput(request_id=0, prompt='explain the fractional banking thoroughly.', prompt_token_ids=[1, 5649, 278, 15958, 284, 9124, 292, 26606, 29889], encoder_prompt=None, encoder_prompt_token_ids=None, prompt_logprobs=None, outputs=[CompletionOutput(index=0, text=' \n<|assistant|>\nFractional banking is a system of banking that allows for the creation of fractional reserve banking, where the bank creates more money than is needed to meet the demands of the economy. In this system, banks create new money out of thin air by', token_ids=[2, 29871, 13, 29966, 29989, 465, 22137, 29989, 29958, 13, 29943, 13857, 284, 9124, 292, 338, 263, 1788, 310, 9124, 292, 393, 6511, 363, 278, 11265, 310, 15958, 284, 23986, 9124, 292, 29892, 988, 278, 9124, 10017, 901, 6909, 1135, 338, 4312, 304, 5870, 278, 1261, 4167, 310, 278, 26504, 29889, 512, 445, 1788, 29892, 24388, 1653, 716, 6909, 714, 310, 16835, 4799, 491], cumulative_logprob=None, logprobs=None, finish_reason=length, stop_reason=None)], finished=True, metrics=None, lora_request=None, num_cached_tokens=0, multi_modal_placeholders={})

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by elaineyz** [comment]
> Thanks @elaineyz!
> 
> Actually the semantics for `min_tokens` are meant to be that EOS/stop tokens won't be sampled/output until min_tokens have been generated. Which suggests some other bug here. Could you share the parameters you're using to run the server? There is a known issue where min_tokens may not work properly in conjunction with speculative decoding.
> 
> There's a related `--ignore-eos` arg which still samples EOS tokens but doesn't stop on them, though this applies over the entire output.

Hi @njhill, for Neuron hardware we do on device sampling by default, which means the model runner will return token IDs back to the scheduler. In V0 this worked properly since stop checking only happens in 1 place during engine output processing: https://github.com/vllm-project/vllm/blob/v0.9.2/vllm/engine/output_processor/stop_checker.py#L44-L47. 

Do you see any issues with additionally adding this min_tokens check into the V1 scheduler stopping logic? Thanks for your help!

**Comment by njhill** [comment]
@elaineyz I guess my point is that this shouldn't be necessary since the logits should be modified prior to sampling in the min_tokens case such that the eos token won't be sampled (until min_token tokens have been generated).

This is done by the [`MinTokensLogitsProcessor`](https://github.com/vllm-project/vllm/blob/c50901f3b96418c600d4dc20188360cd7a7b1145/vllm/v1/sample/logits_processor/builtin.py#L166). Is there some reason that this wouldn't be invoked in your case? Again it's a known issue that this doesn't currently work with speculative decoding, because the rejection sampler doesn't call the logits processors.

File: vllm/v1/core/sched/utils.py [hunk 1]
```diff
@@ -58,6 +58,11 @@ def check_stop(
 
     sampling_params = request.sampling_params
     assert sampling_params is not None
+
+    min_tokens = sampling_params.min_tokens
+    if min_tokens > 0 and request.num_output_tokens < min_tokens:
```
Comment (njhill) [inline vllm/v1/core/sched/utils.py]: I don't think the first check is necessary?
```suggestion
    if request.num_output_tokens < min_tokens:
```

**Review by njhill — COMMENTED** [review]
Thsanks @elaineyz that makes sense. I think it's fine to add this.

File: vllm/v1/core/sched/utils.py (hunk 1 above)
Comment (elaineyz) [reply vllm/v1/core/sched/utils.py]: sounds good, removed.

**Comment by elaineyz** [comment]
Thanks @njhill, could you trigger the ci check if you are ok with the change?

**Review by njhill — APPROVED** [review]
Thanks @elaineyz
